src/binding/c/datatype_api.txt

```diff
@@ -0,0 +1,542 @@
+    /* handle overflow: see MPI-3 p.104 */
+    *size = (size_x > INT_MAX) ? MPI_UNDEFINED : (int) size_x;
+}
+
+MPI_Type_size_x:
+    .desc: Return the number of bytes occupied by entries in the datatype
+    .skip: initcheck
+
+MPI_Type_set_name:
+    .desc: set datatype name
+    .skip: ThreadSafe
+{
+    /* Include the null in MPI_MAX_OBJECT_NAME */
+    MPL_strncpy(datatype_ptr->name, type_name, MPI_MAX_OBJECT_NAME);
+}
+
+MPI_Type_get_name:
+    .desc: Get the print name for a datatype
+    .docnotes: ThreadSafeNoUpdate, NULL
+{
+    /* Include the null in MPI_MAX_OBJECT_NAME */
+    MPL_strncpy(type_name, datatype_ptr->name, MPI_MAX_OBJECT_NAME);
+    *resultlen = (int) strlen(type_name);
+}
+
+MPI_Type_contiguous:
+    .desc: Creates a contiguous datatype
+
+MPI_Type_create_darray:
+    .desc: Create a datatype representing a distributed array
+    .skip: validate-ORDER, validate-RANK_NNI
+{ -- error_check --
+    CHECKENUM: order, storageorder, MPI_ORDER_C MPI_ORDER_FORTRAN
+    MPIR_ERR_CHKANDJUMP3((rank < 0 || rank >= size), mpi_errno, MPI_ERR_RANK,
+                          "**argrange", "**argrange %s %d %d", "rank", rank, (size - 1));
+}
+
+MPI_Type_create_subarray:
+    .desc: Create a datatype for a subarray of a regular, multidimensional array
+    .skip: validate-ORDER
+{ -- error_check --
+    CHECKENUM: order, storageorder, MPI_ORDER_C MPI_ORDER_FORTRAN
+}
+
+MPI_Type_create_hindexed:
+    .desc: Create a datatype for an indexed datatype with displacements in bytes
+
+MPI_Type_create_hindexed_block:
+    .desc: Create an hindexed datatype with constant-sized blocks
+
+MPI_Type_create_hvector:
+    .desc: Create a datatype with a constant stride given in bytes
+
+MPI_Type_create_indexed_block:
+    .desc: Create an indexed datatype with constant-sized blocks
+/* -- notes-2 --
+    Notes:
+    The indices are displacements, and are based on a zero origin.  A common error
+    is to do something like the following
+    .vb
+        integer a(100)
+        integer blens(10), indices(10)
+        do i=1,10
+    10       indices(i) = 1 + (i-1)*10
+        call MPI_TYPE_CREATE_INDEXED_BLOCK(10,1,indices,MPI_INTEGER,newtype,ierr)
+        call MPI_TYPE_COMMIT(newtype,ierr)
+        call MPI_SEND(a,1,newtype,...)
+    .ve
+    expecting this to send "a(1),a(11),..." because the indices have values
+    "1,11,...".   Because these are `displacements` from the beginning of "a",
+    it actually sends "a(1+1),a(1+11),...".
+    
+    If you wish to consider the displacements as indices into a Fortran array,
+    consider declaring the Fortran array with a zero origin
+    .vb
+        integer a(0:99)
+    .ve
+*/
+
+MPI_Type_create_resized:
+    .desc: Create a datatype with a new lower bound and extent from an existing datatype
+
+MPI_Type_create_struct:
+    .desc: Create an MPI datatype from a general set of datatypes, displacements, and block sizes
+
+MPI_Type_dup:
+    .desc: Duplicate a datatype
+
+MPI_Type_hindexed:
```

Comment:
missing `MPI_ERR_TYPE, MPI_ERR_COUNT, MPI_ERR_ARG`

Reply:
Let's ignore the removed functions